### ISSUE TITLE
Pytorch Simple Example using Pytorch docs training loop

### ISSUE THREAD

**Issue by umairkhawaja** [optuna/optuna-examples#34]
I modified the pytorch simple code example given in this repo and replaced the training/validation loop with a code similar to the sample training loop given in a classifier training example in the pytorch docs. However, the study ends after a single trial. I'm assuming this has something to do with the `return accuracy` statement at the end of the objective function. Can someone explain the purpose of this statement, or better if someone has any idea what I could be doing wrong?

**Comment by nzw0301** [comment]
I'm not sure which PyTorch docs code you modify, but I suppose the objective function is something like this.

```python
def objective(trial):
    # sample hyper-parameters using optuna
    for epoch in range(epochs):
        train_loop()
        val_accuracy = validation()
        # call pruning function 
    return val_accuracy # best_val_accuracy, or val_loss
```

**Comment by umairkhawaja** [comment]
Oh I didn't know it is supposed to be the best accuracy. I was storing the most recent accuracy. Here's my code snippet:

`            if phase == 'val':
                df,_,_ = evaluate_model(model,dataloaders['val'],writer=writer,iteration=epoch,phase='val')
                record = [epoch,'val'] + df['Value'].tolist()
                summary.loc[len(summary)+1] = record
                if epoch_acc >= best_accuracy:
                    best_accuracy = epoch_acc
                trial.report(epoch_acc, epoch)
                # Handle pruning based on the intermediate value.
                if trial.should_prune():
                    raise optuna.exceptions.TrialPruned()
    summary.to_csv(join(logdir,'metrics_summary.csv'),index=False)
    return best_accuracy`

**Comment by nzw0301** [comment]
Well, it depends on what you want. Personally, I used it because I would like to maximise the validation accuracy rather than the validation accuracy at the final epoch. In addition, the accuracy is affected by hyper-parameters such as learning rate or learning rate schedule, etc. Thus I'll save the checkpoint (weights of the model) as well when the best accuracy improves.

Looks great!

**Comment by umairkhawaja** [comment]
That's a really good suggestion, thanks!

I ran the code again, the study concludes after a single trial again. Any ideas what else I could be doing wrong?

**Comment by nzw0301** [comment]
Could you give a minimal reproducible code?

**Comment by umairkhawaja** [comment]
I can't share any version of the code I'm running  as I'm not sure if I'm allowed to but the training loop is almost exactly the same as this one

https://pytorch.org/tutorials/beginner/transfer_learning_tutorial.html

**Comment by nzw0301** [comment]
Okay. In my mind, there are two cases if there is no error and the objective contains hyper-parameter suggestions correctly,

1. `n_trials` of `study.optimize` is `1`.
2. Pytorch's data loader is not working correctly. 

What do you think about it?

**Comment by umairkhawaja** [comment]
My first thought was about `n_trials` but it is set to 100. And I sample hyperparameters like this:
```
lr = trial.suggest_float("lr", 1e-5, 1e-1, log=True)
decay = trial.suggest_float("weight_decay", 1e-5, 5e-1, log=True)
momentum = trial.suggest_float("momentum", 0.7, 0.999, log=True)
```

Any such issue should terminate the trial, no?

**Comment by nzw0301** [comment]
How about `timeout`  of `study.optimize`? If the value is too small, it may stops after 1 trial.

**Comment by umairkhawaja** [comment]
Do you have any recommendations for it?  Currently, it is set to 600

**Comment by nzw0301** [comment]
For now, I suggest using `None` for debugging because the study performs just `n_trials` times.

**Comment by umairkhawaja** [comment]
Thanks! Is there some documentation where I can see the overall flow of it?

**Comment by nzw0301** [comment]
👍  Yes, I think the explanations of the [`study.optimize`'s arguments](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.study.Study.html#optuna.study.Study.optimize).

**Comment by nzw0301** [comment]
Has your issue been solved by not specifying `timeout`?

**Comment by umairkhawaja** [comment]
> Has your issue been solved by not specifying `timeout`?

Oh yes, specifying None worked. Apologies for not updating here. Thank you!

**Comment by nzw0301** [comment]
Nice 👍 Can we close this issue?

**Comment by umairkhawaja** [comment]
> Nice 👍 Can we close this issue?

Yes, sure
